File: kids/frontend/SwiftUI/Weather/viewModel/WeatherViewModel.swift
//
//  WeatherViewModel.swift
//  Weather
//
//  Created by Wenwei Tao on 5/23/20.
//  Copyright © 2020 Wenwei Tao. All rights reserved.
//

import Foundation

let cities = ["Shanghai", "London", "San Francisco"]

struct WeatherModel: Identifiable {
    let id: Int
    let city: String
    let temperature: Double
}

class WeatherViewModel: ObservableObject {
     @Published var weathers = [WeatherModel]()
    
    func loadData() {
        var i = 0
        for city in cities {
            WeatherApi.getCurrentTemperature(city) { weather in
                print(weather.the_temp)
                self.weathers.append(WeatherModel(id: i, city: city, temperature: weather.the_temp))
                i = i + 1
            }
        }
    }
}

File: kids/frontend/react.js/src/viewModel/WeatherViewModel.js
import React, { useState, createContext, useEffect } from "react";
import WeatherApi from "../model/WeatherApi";

// Create Context Object
export const WeatherContext = createContext();

// Create a provider for components to consume and subscribe to changes
export const WeatherProvider = (props) => {
    const [weathers, setWeathers] = useState([]);

    useEffect(() => {
        const cities = ["Shanghai", "London", "San Francisco"];
        const loadData = async () => {
            let items = [];
            for (const cityName of cities) {
                let cityWeather = await WeatherApi.getCurrentTemperature(
                    cityName
                );
                items = [
                    {
                        city: cityName,
                        temperature:
                            cityWeather.consolidated_weather[0].the_temp,
                    },
                    ...items,
                ];
            }
            setWeathers(items);
        };
        loadData();
    }, []);

    return (
        <WeatherContext.Provider value={weathers}>
            {props.children}
        </WeatherContext.Provider>
    );
};
